docs/bounded_dynamism.md

```diff
@@ -0,0 +1,85 @@
+
+
+## (P3) get\_dimension\_size / set\_dimension\_size ops
+
+The `get_dimension_size` op takes a tensor and dimension index as operands and returns the actual size of the dimension at runtime as `i32` type scalar.
+
+The following example returns the size of the result after concatenating input that has up to `16` elements with self and returns the actual runtime size of the concatenation result.
+
+
+```
+func.func @self_concat_size(%data: tensor<?xi32, #stablehlo.type_extensions<bounds = [16]>>) -> tensor<i32> {
+  %concat = "stablehlo.concatenate"(%data, %data) {dimension = 0 : i64}
+    : (tensor<?xi32, #stablehlo.type_extensions<bounds = [16]>>, 
+       tensor<?xi32, #stablehlo.type_extensions<bounds = [16]>>)
+      -> tensor<?xi32, #stablehlo.type_extensions<bounds = [32]>>
+
+  %result = "stablehlo.get_dimension_size"(%concat) {dimension = 0 : i64}
+    : (tensor<?xi32, #stablehlo.type_extensions<bounds = [32]>>) -> tensor<i32>
+  
+  func.return %result : tensor<i32>
+}
+```
+
+
+The `set_dimension_size` op takes a tensor, runtime size and dimension index as operands and returns a tensor whose logical size of the particular dimension is set to the specified value. This size needs to be less than or equal to the static size, if available. 
+
+In the following example, set dimension size is used to set the logical size of the first dimension. With input `[1, 2, 3, 4]`, it returns `3` with batch\_size equal to `2` but returns `6` with batch\_size `3`.
+
+
+```
+func.func @dynamic_sum(%data: tensor<4xi32>, %batch_size: tensor<i32>) -> tensor<i32> {
+  %dim = stablehlo.constant dense<0> : tensor<i32>
+  %dynamic_data =  "stablehlo.set_dimension_size"(%data, %batch_size, %dim)
+    : (tensor<4xi32>, tensor<i32>, tensor<i32>) 
+      -> tensor<?xi32, #stablehlo.type_extensions<bounds = [4]>>
+  
+  %zero = stablehlo.constant dense<0> : tensor<i32>
+  %sum = "stablehlo.reduce"(%dynamic_data, %zero) ({
+   ^bb0(%arg1: tensor<i32>, %arg2: tensor<i32>):
+    %add = stablehlo.add %arg1, %arg2 : tensor<i32>
+    "stablehlo.return"(%add) : (tensor<i32>) -> ()
+  }) {dimensions = dense<[0]> : tensor<1xi64>}
+    : (tensor<?xi32, #stablehlo.type_extensions<bounds = [4]>>, 
+       tensor<i32>) -> tensor<i32>
+  func.return %sum : tensor<i32>
+}
+```
+
+
+Currently, `get_dimension_size` result` and set_dimension_size `size operand only supports `i32` type for legacy reasons. This is not consistent with other types for dimension sizes so it should be expanded to also allow `i64` typed operand for `size`. The default result type for `get_dimension_size` should be `i64` in the shape inference functions but allows `i32` types as well.
+
+
+## (P4) Prefer generic dynamic ops over set\_dimension\_size op
```

Comment:
This is a promising idea, but at the moment it's not worked out to the level of detail that allows to easily evaluate it for potential StableHLO users.

In an RFC like this one, I think that it is important to provide a clear recommendation for what IR should be produced by frontends and consumed by compilers to achieve a certain goal, i.e. express bounded computations in this case. 

I think that (P1)-(P3) are pretty clear about this. To express a bounded computation, use unbounded computations with tensors that have bounds. To set bounds, either pass already bounded tensors as arguments or use `set_dimension_size`. That's as clear as it gets while we're waiting for the unbounded dynamism RFC, which will clarify which ops can be used to express unbounded computations.

In comparison, (P4) feels handwavy because: 
  1. It doesn't explain how to express bounded computations in a new way (intuitively, I think some kind of a slice will suffice, but this needs at least an example that works out the details).
  1. It says that the new idea doesn't apply to all use cases, but doesn't survey how important they are.
  1. It mentions a transformation from unbounded dynamism to bounded dynamism as the cornerstone of the new idea, but doesn't explain how this transformation works. This detail is important to know for: a) producers who want to reliably express bounded computations in the new way, b) consumers who want to support bounded computations on hardware which doesn't support unbounded dynamism.

Finally, (P4) doesn't provide a motivation for why having two ways of expressing bounded computations (via `set_dimension_size` and via `slice`) is better than have just one way (via `set_dimension_size`). I think the motivation would be clear if it was proposed to supersede `set_dimension_size` with `slice`, but the current formulation of (P4) doesn't look like an obvious win.

Reply:
As a consumer of StableHLO, PyTorch/XLA has direct utility for the ops mentioned in this section. It's unclear how the path forward will look like without these ops. Let's collaborate on a solution that lands a concrete and east-to-adopt replacement.

Reference to the API calls that PyTorch/XLA calls in its lowering stack: https://github.com/pytorch/xla/blob/e0eb2a7bbc1e7dcefa4d83c88548a9dad7c3a387/torch_xla/csrc/xla_op_builder.cpp#L684-L695